### PR TITLE
[FLINK-24689][runtime-web] Add log's last modify time in log list view

### PR DESCRIPTION
**Pull request by ysymi**
## What is the purpose of the change

This pull request add log's **last modified time** in taskmanager log list view and jobmanager log list view, it will help us to choose correct log file and analyse problem exactly and quickly. 

See [FLINK-24689](https://issues.apache.org/jira/browse/FLINK-24689) for more.

## Brief change log

  - add `mtime` field in class `LogInfo` and put log file's last modified time into rest api's response
  - show `Last Modified Time` in TaskManager's log list view and JobManager's log list view (support sort)

## Verifying this change

This change is already covered by existing tests:
- org.apache.flink.runtime.rest.handler.cluster.JobManagerLogListHandlerTest#testGetJobManagerLogsList
- org.apache.flink.runtime.rest.handler.taskmanager.TaskManagerLogListHandlerTest#testGetTaskManagerLogsList
- org.apache.flink.runtime.rest.messages.taskmanager.LogListInfoTest#getTestResponseInstance

I filled mtime's value and pass these testcase.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no 
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  no

### CODE REVIEW

File: flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html
Comment (Airblader) [inline flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html:37]: What type is inferred for `table.data`? If this is just inferred to be `any`, then this would be a regression in type safety.

Comment (ysymi) [reply flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html:37]: Thanks for review code firstly.  
I think table.data's type here is `TaskManagerLogItemInterface[]` and table.data's type in `job-manager-log-list.component.html` is `JobManagerLogItemInterface[]`. 
I read the doc of [NG_ZORRO#table](https://ng.ant.design/components/table/en#components-table-demo-basic) , and it's said that table.data's value is set by nz-table's property `[nzData]`, so I passed `listOfLog` to `nzData` which is defined in related .ts file .

**Comment by ysymi** [comment]
@flinkbot re-run azure

**Comment by ysymi** [comment]
I will rebase on latest master (commit: fc4f2556) to pass the CI check in Azure.
The reason is the master I checkout from (commit: db04af41 ) always failed in `org.apache.flink.streaming.connectors.kafka.table.KafkaDynamicTableFactoryTest#testTableSourceWithKeyValueAndMetadata`, because the valueDecodingFormat.producedDataType was different between two KafkaSource(one has `not null` but another one not have). And looks like this problem has been solved by commit 2a582f3. 
So I will rebase on laster master to pass the CI

File: flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html
Comment (Airblader) [reply flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html:37]: I don't mean what the type "should" be, but rather what the Angular Language Service will _actually_ infer. The last time I checked, Angular cannot use component generics to infer the type correctly. You can hover over the `log` in `let log` and see the type the language service will infer. If this is `any`, that's not good.

Comment (ysymi) [reply flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html:37]: I got it and I found that the type of `log` in  `let log in table.data` is `unknown` and  type of `table.data` is `unknown[]` by the way you said in IDE. I should change it to `let log in listOfLog` and then type of `log` will be clear and correct, am I right ? @Airblader

Comment (Airblader) [reply flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html:37]: If the change to `table.data` isn't needed, I would prefer leaving it as `listOfLog` for now, yes. If we need the change (like for pagination) then we should at least use the same type trick I mentioned [here](https://github.com/apache/flink/pull/17619#discussion_r741033701).

Comment (ysymi) [reply flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html:37]: I found that change to `listOfLog` will lead sortFn not work, and if I add `trackBy` function in ngFor loop, the type of `log` will be clear and correct. I'm not familiar with Angular, so I'm not sure if it's a type trick or not. However add trackBy function seems like a way that works.

Comment (ysymi) [reply flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html:37]: @Airblader how do you think of add trackBy function to help angular infer correct type ?

Comment (Airblader) [reply flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html:37]: `trackBy` is a useful addition regardless; if it happens to make the type inference work better as well, then that's fine with me.

Comment (ysymi) [reply flink-runtime-web/web-dashboard/src/app/pages/task-manager/log-list/task-manager-log-list.component.html]: @Airblader I added trackBy function and this pr is ready to be reviewed again.
And please help me to rerun the CI , latest CI check was failed cause by flink-connector-rabbitmq 's integration-tests which is occasionally I think, thanks !

**Review by Airblader — APPROVED** [review]
The changes LGTM. For the CI you can try to rebase to the current master branch and it will run again, maybe the issue was addressed already.

**Comment by ysymi** [comment]
CI check passed again finally after rebase on lastest master (commit: 059c00f4)

**Comment by ysymi** [comment]
Hi @Airblader , I want to know who will merge this pr ?  You or the committer who assigned jira ticket to me?

**Comment by Airblader** [comment]
I'm not a committer, so I don't have the rights to merge this. You'll need to find a committer who can, maybe @tillrohrmann can help with that since he assigned you.

**Comment by ysymi** [comment]
Thank you very much, @Airblader. 😁
And @tillrohrmann, please have a look for this pr, if there is anything I need to continue with, just let me know.

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @ysymi and the thorough review @Airblader. The changes look good to me. Could you rebase your PR one last time? There are some conflicting changes. Once it is rebased and there are no more merge conflicts, I will merge this PR.

**Comment by ysymi** [comment]
@tillrohrmann I have rebased on latest master (commit: 3262d9b) and add a commit for fixing a coding-style problem

**Comment by tillrohrmann** [comment]
Thanks a lot @ysymi. Merging this PR now.

**Pull request closed without merging**
